deepinv/training_utils/trainer.py

```diff
@@ -0,0 +1,714 @@
+        :param torch.Tensor x: Ground truth.
+        :param torch.Tensor y: Measurement.
+        :param torch.Tensor x_net: Network reconstruction.
+        :param bool train: If ``True``, the model is trained, otherwise it is evaluated.
+        """
+        post_str = "Training" if train else "Eval"
+        if self.plot_images and ((epoch + 1) % self.freq_plot == 0):
+            imgs, titles, grid_image, caption = self.prepare_images(
+                physics, x, y, x_net
+            )
+
+            plot(
+                imgs,
+                titles=titles,
+                show=self.plot_images,
+                return_fig=True,
+                rescale_mode="clip",
+            )
+
+            if self.wandb_vis:
+                log_dict_post_epoch = {}
+                images = wandb.Image(
+                    grid_image,
+                    caption=caption,
+                )
+                log_dict_post_epoch[post_str + " samples"] = images
+                log_dict_post_epoch["step"] = epoch
+                wandb.log(log_dict_post_epoch)
+
+    def save_model(self, epoch, eval_psnr=None):
+        r"""
+        Save the model.
+
+        It saves the model every ``ckp_interval`` epochs.
+
+        :param int epoch: Current epoch.
+        :param None, float eval_psnr: Evaluation PSNR.
+        """
+        if (epoch > 0 and epoch % self.ckp_interval == 0) or epoch + 1 == self.epochs:
+            os.makedirs(str(self.save_path), exist_ok=True)
+            state = {
+                "epoch": epoch,
+                "state_dict": self.model.state_dict(),
+                "loss": self.loss_history,
+                "optimizer": self.optimizer.state_dict(),
+            }
+            if eval_psnr is not None:
+                state["eval_psnr"] = eval_psnr
+            torch.save(
+                state, os.path.join(str(self.save_path), "ckp_{}.pth.tar".format(epoch))
+            )
+
+    def train(
+        self,
+    ):
+        r"""
+        Train the model.
+
+        It performs the training process, including the setup, the evaluation, the forward and backward passes,
+        and the visualization.
+
+        :returns: The trained model.
+        """
+
+        self.setup_train()
+
+        for epoch in range(self.epoch_start, self.epochs):
+            ## Evaluation
+            perform_eval = self.eval_dataloader and (
+                (epoch + 1) % self.eval_interval == 0 or epoch + 1 == self.epochs
+            )
+            if perform_eval:
+                self.current_iterators = [
+                    iter(loader) for loader in self.eval_dataloader
+                ]
+                batches = len(self.eval_dataloader[self.G - 1]) - int(
+                    self.eval_dataloader[self.G - 1].drop_last
+                )
+
+                self.model.eval()
+                for i in (
+                    progress_bar := tqdm(
+                        range(batches), ncols=150, disable=not self.verbose
```

Comment:
Can we have some more fine-grained verbosity / a param for `progress_bar=True` ? E.g. in sphinx examples we don't want all the `tqdm` stdout, but do want metric printouts (see e.g. https://deepinv.github.io/deepinv/auto_examples/self-supervised-learning/demo_equivariant_imaging.html#train-the-network)

Reply:
One solution is to have in `step()`:
```python
if last_batch:
    if verbose and not progress_bar:
        print(f"{'Train' if train else 'Eval'} epoch {epoch}: {' '.join([f'{k}={round(v, 3)}' for (k, v) in logs.items()])}")
```
i.e. just print formatted logs when progress_bar is off. And then `tqdm` also has `disable=not self.verbose and not progress_bar`

Reply:
done